Web Inspector: Network: allow the method of Request Local Overrides to passthrough

dcrousso opened the pull request:
#### 8b9a751017aa0d969a219404b0d59def58c92acb
<pre>
Web Inspector: Network: allow the method of Request Local Overrides to passthrough
<a href="https://bugs.webkit.org/show_bug.cgi?id=282076">https://bugs.webkit.org/show_bug.cgi?id=282076</a>

Reviewed by Anne van Kesteren.

This allows developers to use a Request Local Override to add headers to all requests matching the given URL without modifying any of the other data.

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.resourceRequestWasServedFromMemoryCache):
(WI.NetworkManager.prototype.requestIntercepted):

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
(WI.LocalResourceOverridePopover.prototype.get serializedData):
(WI.LocalResourceOverridePopover.prototype.show):
Add a passthrough option to the method select when passthrough is enabled.

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverrideRequestContentView.js:
(WI.LocalResourceOverrideRequestContentView.prototype.initialLayout):
Only show the message about the request not having a body if the request method exists and is one of those.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource):
* Source/WebInspectorUI/UserInterface/Models/WebSocketResource.js:
(WI.WebSocketResource):
Drive-by: use `WI.HTTPUtilities.RequestMethod.GET` instead of `&quot;GET&quot;`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/http/tests/inspector/network/intercept-request-properties.html:

Canonical link: <a href="https://commits.webkit.org/286679@main">https://commits.webkit.org/286679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a860aa2604a8672b3c5c8a3eee2a81d4b9e56c3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77806 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24752 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62074 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/728 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57782 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16196 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76581 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63253 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38192 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44470 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20739 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23083 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/830 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66162 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/971 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65440 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16875 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9298 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7476 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/793 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/823 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/809 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -992,19 +992,20 @@ WI.NetworkManager = class NetworkManager extends WI.Object
                 return;
 
             case WI.LocalResourceOverride.InterceptType.Request: {
+                let method = localResource.requestMethod ?? (isPassthrough ? request.method : "");
```
Why the empty string?

dcrousso replied:
if the developer does not provide a HTTP method and has not enabled passthrough then we will interpret that as "remove the HTTP method" (just like how if no HTTP headers are provided and passthrough is not enabled then that's like saying "remove all headers")

the reason we have the passthrough option is to allow developers to differentiate between "please remove this value" and "please use whatever this value would be if the local override didn't exist"

annevk replied:
I guess I was wondering why this wasn't null or some such. But if empty string has that meaning, so be it.

annevk commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -992,19 +992,20 @@ WI.NetworkManager = class NetworkManager extends WI.Object
-                        if (isPassthrough)
-                            return request.data;
-                        return "";
+                        if (method && WI.HTTPUtilities.RequestMethodsWithBody.has(method)) {
```
Wouldn't returning early here be possible and also be the preferred style?

dcrousso replied:
i didn't like the idea of having 2 different `return undefined;`

annevk reviewed (commented):
It would be great to have some test coverage for this change.

dcrousso committed:
Web Inspector: Network: allow the method of Request Local Overrides to passthrough
https://bugs.webkit.org/show_bug.cgi?id=282076

Reviewed by Anne van Kesteren.

This allows developers to use a Request Local Override to add headers to all requests matching the given URL without modifying any of the other data.

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.resourceRequestWasServedFromMemoryCache):
(WI.NetworkManager.prototype.requestIntercepted):

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
(WI.LocalResourceOverridePopover.prototype.get serializedData):
(WI.LocalResourceOverridePopover.prototype.show):
Add a passthrough option to the method select when passthrough is enabled.

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverrideRequestContentView.js:
(WI.LocalResourceOverrideRequestContentView.prototype.initialLayout):
Only show the message about the request not having a body if the request method exists and is one of those.

* Source/WebInspectorUI/UserInterface/Models/Resource.js:
(WI.Resource):
* Source/WebInspectorUI/UserInterface/Models/WebSocketResource.js:
(WI.WebSocketResource):
Drive-by: use `WI.HTTPUtilities.RequestMethod.GET` instead of `"GET"`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/http/tests/inspector/network/intercept-request-properties.html:

Canonical link: https://commits.webkit.org/286679@main
